File: src/main/java/sharev/domain/member/controller/MemberController.java
package sharev.domain.member.controller;

import jakarta.validation.Valid;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import sharev.domain.account.entity.Account;
import sharev.domain.member.dto.request.RequestInviteMemberDto;
import sharev.domain.member.dto.request.RequestUpdateMemberRoleDto;
import sharev.domain.member.dto.response.ResponseMemberDto;
import sharev.domain.member.service.MemberService;

@RestController
@RequestMapping("/teams/{teamId}/members")
@RequiredArgsConstructor
public class MemberController {
    private final MemberService memberService;

    @GetMapping
    @PreAuthorize("@teamService.isMember(authentication.principal, #teamId)")
    public ResponseEntity<List<ResponseMemberDto>> getMembers(@PathVariable Long teamId) {
        return ResponseEntity.ok(memberService.getMembers(teamId));
    }

    @PostMapping
    @PreAuthorize("@memberService.isAdmin(authentication.principal, #teamId)")
    public ResponseEntity<Void> invite(@AuthenticationPrincipal Account account,
                                       @PathVariable Long teamId,
                                       @Valid @RequestBody RequestInviteMemberDto requestInviteMemberDto) {
        // TODO: 헥사고날로 변경 후 어떤 식으로 초대 요청 보낼지 고민할 것(이메일도 임시, 로직 설명만을 위함)
        memberService.invite(account, teamId, requestInviteMemberDto.email());
        return ResponseEntity.status(HttpStatus.CREATED).build();
    }

    @PatchMapping("/me/accept")
    @PreAuthorize("@teamService.isMember(authentication.principal, #teamId)")
    public ResponseEntity<Void> acceptInvitation(@AuthenticationPrincipal Account account,
                                                 @PathVariable Long teamId) {
        memberService.acceptInvitation(account, teamId);
        return ResponseEntity.ok().build();
    }

    @DeleteMapping("/me")
    @PreAuthorize("@teamService.isMember(authentication.principal, #teamId)")
    public ResponseEntity<Void> leave(@AuthenticationPrincipal Account account,
                                      @PathVariable Long teamId) {
        memberService.leave(account, teamId);
        return ResponseEntity.noContent().build();
    }

    @PatchMapping("/{memberId}/role")
    @PreAuthorize("@memberService.isAdmin(authentication.principal, #teamId)")
    public ResponseEntity<Void> updateRole(@PathVariable Long teamId,
                                           @PathVariable Long memberId,
                                           @Valid @RequestBody RequestUpdateMemberRoleDto requestUpdateMemberRoleDto) {
        memberService.updateRole(teamId, memberId, requestUpdateMemberRoleDto.role());
        return ResponseEntity.ok().build();
    }

    @DeleteMapping("/{memberId}")
    @PreAuthorize("@memberService.isAdmin(authentication.principal, #teamId)")
    public ResponseEntity<Void> removeMember(@AuthenticationPrincipal Account account,
                                             @PathVariable Long teamId,
                                             @PathVariable Long memberId) {
        memberService.removeMember(account, teamId, memberId);
        return ResponseEntity.noContent().build();
    }
}

File: src/main/kotlin/sharev/member/adapter/inbound/web/controller/MemberController.kt
package sharev.member.adapter.inbound.web.controller

import jakarta.validation.Valid
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*
import sharev.common.adapter.inbound.security.model.AccountPrincipal
import sharev.member.adapter.inbound.web.dto.request.InviteMemberRequest
import sharev.member.adapter.inbound.web.dto.request.UpdateMemberRoleRequest
import sharev.member.adapter.inbound.web.dto.response.*
import sharev.member.adapter.inbound.web.mapper.toCommand
import sharev.member.adapter.inbound.web.mapper.toResponse
import sharev.member.application.port.inbound.command.AcceptInvitationCommand
import sharev.member.application.port.inbound.command.GetMembersCommand
import sharev.member.application.port.inbound.command.LeaveTeamCommand
import sharev.member.application.port.inbound.command.RemoveMemberCommand
import sharev.member.application.port.inbound.usecase.*

@RestController
@RequestMapping("/teams/{teamId}/members")
class MemberController(
    private val getMembersUseCase: GetMembersUseCase,
    private val inviteMemberUseCase: InviteMemberUseCase,
    private val acceptInvitationUseCase: AcceptInvitationUseCase,
    private val leaveTeamUseCase: LeaveTeamUseCase,
    private val updateMemberRoleUseCase: UpdateMemberRoleUseCase,
    private val removeMemberUseCase: RemoveMemberUseCase,
) {

    @GetMapping
    fun getMembers(
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
        @PathVariable teamId: Long,
    ): ResponseEntity<List<MemberResponse>> {
        val response = getMembersUseCase.getMembers(
            GetMembersCommand(accountPrincipal.id, teamId)
        ).map { it.toResponse() }

        return ResponseEntity.ok(response)
    }

    @PostMapping
    fun invite(
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
        @PathVariable teamId: Long,
        @Valid @RequestBody request: InviteMemberRequest,
    ): ResponseEntity<InviteMemberResponse> {
        val response = inviteMemberUseCase.invite(request.toCommand(accountPrincipal, teamId))
            .toResponse()

        return ResponseEntity.ok(response)
    }

    @PatchMapping("/me/accept")
    fun acceptInvitation(
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
        @PathVariable teamId: Long,
    ): ResponseEntity<AcceptInvitationResponse> {
        val response = acceptInvitationUseCase.acceptInvitation(
            AcceptInvitationCommand(accountPrincipal.id, teamId)
        ).toResponse()

        return ResponseEntity.ok(response)
    }

    @DeleteMapping("/me")
    fun leave(
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
        @PathVariable teamId: Long,
    ): ResponseEntity<LeaveTeamResponse> {
        val response = leaveTeamUseCase.leave(
            LeaveTeamCommand(accountPrincipal.id, teamId)
        ).toResponse()

        return ResponseEntity.ok(response)
    }

    @PatchMapping("/{memberId}/role")
    fun updateRole(
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
        @PathVariable teamId: Long,
        @PathVariable memberId: Long,
        @Valid @RequestBody request: UpdateMemberRoleRequest,
    ): ResponseEntity<UpdateMemberRoleResponse> {
        val response = updateMemberRoleUseCase.updateRole(
            request.toCommand(accountPrincipal, teamId, memberId)
        ).toResponse()

        return ResponseEntity.ok(response)
    }

    @DeleteMapping("/{memberId}")
    fun removeMember(
        @AuthenticationPrincipal accountPrincipal: AccountPrincipal,
        @PathVariable teamId: Long,
        @PathVariable memberId: Long,
    ): ResponseEntity<RemoveMemberResponse> {
        val response = removeMemberUseCase.removeMember(
            RemoveMemberCommand(accountPrincipal.id, teamId, memberId)
        ).toResponse()

        return ResponseEntity.ok(response)
    }
}
